src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+    } else {
+      this.fileSize = size;
+      LOGGER.info("set buffer size limitation to {} bytes", size);
+    }
+  }
+
+  @Override
+  public void setTopic2TableMap(Map<String, String> topic2TableMap) {
+    this.topic2TableMap = topic2TableMap;
+  }
+
+  @Override
+  public void setFlushTime(long time) {
+    if (time < SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_MIN) {
+      LOGGER.error(
+          "flush time is {} seconds, it is smaller than the minimum "
+              + "flush time {} seconds, reset to the minimum flush time",
+          time,
+          SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_MIN);
+      this.flushTime = SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_MIN;
+    } else {
+      this.flushTime = time;
+      LOGGER.info("set flush time to {} seconds", time);
+    }
+  }
+
+  @Override
+  public void setMetadataConfig(SnowflakeMetadataConfig configMap) {
+    this.recordService.setMetadataConfig(configMap);
+  }
+
+  @Override
+  public long getRecordNumber() {
+    return this.recordNum;
+  }
+
+  @Override
+  public long getFlushTime() {
+    return this.flushTime;
+  }
+
+  @Override
+  public long getFileSize() {
+    return this.fileSize;
+  }
+
+  @Override
+  public void setBehaviorOnNullValuesConfig(
+      SnowflakeSinkConnectorConfig.BehaviorOnNullValues behavior) {
+    this.behaviorOnNullValues = behavior;
+  }
+
+  @Override
+  public void setCustomJMXMetrics(boolean enableJMX) {
+    this.enableCustomJMXMonitoring = enableJMX;
+  }
+
+  @Override
+  public SnowflakeSinkConnectorConfig.BehaviorOnNullValues getBehaviorOnNullValuesConfig() {
+    return this.behaviorOnNullValues;
+  }
+
+  @Override
+  public void setDeliveryGuarantee(
+      SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee ingestionDeliveryGuarantee) {
+    this.ingestionDeliveryGuarantee = ingestionDeliveryGuarantee;
+  }
+
+  @Override
+  public Optional<MetricRegistry> getMetricRegistry(String pipeName) {
+    return Optional.empty();
+  }
+
+  /**
+   * Gets a unique identifier consisting of topic name and partition number.
+   *
+   * @param topic topic name
+   * @param partition partition number
+   * @return combinartion of topic and partition
+   */
+  @VisibleForTesting
+  protected static String partitionChannelKey(String topic, int partition) {
+    return topic + "_" + partition;
+  }
+
+  // ------ Streaming Ingest Related Functions ------ //
+  private OpenChannelRequest getOpenChannelRequest(
+      final String channelName, final String tableName) {
+    return OpenChannelRequest.builder(channelName)
+        .setDBName(this.connectorConfig.get("snowflake.database.name"))
+        .setSchemaName(this.connectorConfig.get("snowflake.schema.name"))
+        .setTableName(tableName)
+        .build();
+  }
+
+  private void closeStreamingClient() {
+    // do we need to close the client? If I close, I will have to re init the client upon rebalance
```

Comment:
good question, if you look at `SnowflakeStreamingIngestClient.close`, it does two things:
1. it will flush all the rows in all the channels, and mark the client as closed so no new inserts are allowed
2. It will check whether any of the channels have uncommitted rows, after flushing

I'm not sure how rebalance works, so I can't comment

Reply:
Thanks Toby. 

So rebalancing is pretty straight forward. It's like a server restart but new partitions can be assigned to same node and current partitions can be given to other node. 

So when a rebalance happens:
1. close() and 
2. open() 
APIS are called in that order. 


Follow up question. 
1. Should I use client close or channel close?
2. Do I need to init channel in open ()? I only do openChannel which is fine for first time because start() will be called but start() wont be called after rebalance. 
  a. So, will openChannel init the client or do I need to invoke isClosed()?

Reply:
1. You could just use client.close if all channels need to be closed, but one question I have is what would you do if there is a failure?
2. You need to reinit the channel if you have close it. Actually, will the `partitionsToChannel` be gone during rebalance? 

I guess it will be easier to schedule a quick call, i'm not sure about your timezone, do you want to schedule a chat on my calendar?

Reply:
I added 2. in my recent commits. 

for 1, I will handle in other PR.(
failure handling)

Reply:
(I realized this while writing tests)

Reply:
Here is what happens during rebalance. 
1. open(Partitions) followed by close(Partitions)
2. open(Partitions) is also called immediately after connect starts. 

We might not have to call close() channel from close(partition) because we will have to reopen them. More on that in further PRs. I will go with this current implementation which is much cleaner although involves few more remote calls.

Reply:
Answering specific questions as this can easily be forgotten. 

> You could just use client.close if all channels need to be closed, but one question I have is what would you do if there is a failure?
I will handle this in failure handling PR. If closing has failed, it should be fine since there is no potential impact. May be reopening wouldn't work since it was open. 

> You need to reinit the channel if you have close it. Actually, will the partitionsToChannel be gone during rebalance?
partitionsToChannel will be gone if close() completed successfully.